[Swift] Add dictionaries concept exercise

wneumann opened the pull request:
Forked from elixir/maps

wneumann commented:
Honestly, for the non-code parts of the PR, comments from someone who doesn't know Swift can be more valuable. They can let me know how well I'm getting the ideas across.

stillleben commented on languages/swift/exercises/concept/loops/README.md:
```diff
@@ -0,0 +1,55 @@
+You are an elf working in Santa Claus' logistics division and you have been given a pair of tasks from the boss to help determine the gifts that Santa will deliver.
```
See comment on `dictionaries/README.md`

stillleben commented on languages/swift/exercises/concept/loops/.docs/instructions.md:
````diff
@@ -0,0 +1,55 @@
+// => [(980035519, 1921828986), (1402277754, 712641739), (1821223877, 2651471804)]
+```
+
+## 2. Compute the ranking level for a hashed ID
````
Yeay, that sounds little challenging to solve. :)
I find the explanation with the 2 given examples understandable.
I'm little bit worried, that given hints in `hints.md` might not be sufficient for students, who want hints. What about giving more guidance by describing the rough implementation idea: 1) a loop checking for the "is digital sum a multiple or number?" constraint (with more explanation, when to exit), which is using 2) an inner loop, which calculates the digital sum of a given number by iterating the digits.

stillleben commented on languages/swift/exercises/concept/loops/.docs/after.md:
```diff
@@ -0,0 +1,267 @@
+## Loops
```
See my comment for the other `after.md` file.
I would keep only additional information here: additional links and the "labels" section (otherwise it is likely that this extra section is overseen by students).

stillleben commented:
By the way, you would do a reviewer a favor if you split 2 exercises into 2 prs :) This was quite huge.

wneumann commented on languages/swift/exercises/concept/dictionaries/.docs/introduction.md:
```diff
@@ -0,0 +1,66 @@
+[Dictionaries][dictionaries] are one of Swift's three primary collection types. Dictionaries store mappings between _keys_ which are elements of one type and _values_ which are elements of another type (possibly the same type as that of the keys.
```
Yeah. I believe I'd seen links in a couple of other exercises so I added them, but I agree that it can lead a student down a rabbit hole. I'll pull them.

wneumann commented on languages/swift/exercises/concept/dictionaries/.docs/introduction.md:
````diff
@@ -0,0 +1,66 @@
+// compiler eror: "Cannot assign through subscript: 'constants' is a 'let' constant"
+```
+
+Like arrays, the _key: value_ pairs of a dictionary can be stepped through one at a time using a for-in loop. This type of loop takes each _key: value_ pair of the dictionary and binds the pair to a specified name for further processing inside the loop body. The pair is presented as a tuple and can be destructured like other tuples to assign each element to a name. For example, to print out the a ...
````
Yeah, destructure is used at times in other places like the Swift forums, but I should be consistent with the language docs and go with decompose.

wneumann commented on languages/swift/exercises/concept/dictionaries/.docs/after.md:
```diff
@@ -0,0 +1,70 @@
+[Dictionaries][dictionaries] are one of Swift's three primary collection types. Dictionaries store mappings between _keys_ which are elements of one type and _values_ which are elements of another type (possibly the same type as that of the keys. Unlike arrays, dictionaries are unordered collections. There is no way to, for example, add an entry to the front or back of the dictionary.
```
This is a bit tricky as some of the new material ties into what was covered in the introduction and would look odd if just dropped in out of the blue. But I will edit the lead-in material down so it isn't so long and it doesn't look like the same thing all over again.

Let me do a rewrite/edit and we can see how it looks.

wneumann commented on languages/swift/exercises/concept/dictionaries/.docs/instructions.md:
```diff
@@ -0,0 +1,92 @@
+
+## 1. Define a new high score dictionary
+
+Create a function that returns a new high score dictionary which uses values of type `String` for the keys and values of type `Int` as the values.
```
Ah, yes. Originally it was just creating a new dictionary rather than writing a function and I missed adding the function name in when I re-wrote it.

wneumann commented on languages/swift/exercises/concept/dictionaries/.meta/design.md:
```diff
@@ -0,0 +1,33 @@
+
+## Prerequisites
+
+- `arrays`
```
I've been told that for the prerequisites, you should leave out prerequisites of prerequisites, so `basics` should be left out here because it is a prerequisite of `arrays`.

wneumann commented on languages/swift/exercises/concept/dictionaries/.docs/instructions.md:
```diff
@@ -0,0 +1,92 @@
+
+To add a player to the high score dictionary, define `addPlayer`, which is a function which takes 3 parameters:
+
+- The first parameter is the dictionary of scores. This should be an in-out parameter.
```
Yes. There is a `functions` concept exercise and I need to add it to the prerequisites of this exercise. Good catch.

wneumann commented on languages/swift/exercises/concept/dictionaries/README.md:
```diff
@@ -0,0 +1,92 @@
+In this exercise, you're implementing a way to keep track of the high scores for the most popular game in your local arcade hall.
```
I do not know if it is needed per se. The README.md file is created automatically by the Swift Package Manager along with the rest of the scaffolding for the exercise code. In the V2 Swift track, they copied the instructions file into the code's README, so I continued that here in V3.

wneumann commented:
How very odd. The loops files were submitted in their own separate PR (https://github.com/exercism/v3/pull/1821). I have no idea how they wound up in this one as well. I will have to pull them out.

wneumann commented:
@stillleben I've pushed an update to the swift dictionaries exercise. The update removed the loops concept exercises that were wrongly included in this PR (though I will be addressing your comments on those files in the [loops PR](https://github.com/exercism/v3/pull/1821)). I also edited the introduction and instructions to address your comments on them, and I completely rewrote the after.md in the stye used [in the elixir track](https://github.com/exercism/v3/blob/master/languages/elixir/exercises/concept/maps/.docs/after.md).

Let me know if you think these address your concerns properly.

stillleben commented on languages/swift/exercises/concept/dictionaries/.docs/after.md:
```diff
@@ -1,32 +1,24 @@
+Excellent work. As [dictionaries][dictionaries] are one of Swift's three primary collection types, knowing how to work with dictionaries is a big part of knowing how to program in Swift.
 
-Dictionary literals are written as a series of `key: value` pairs, separated by commas, enclosed in square brackets. Empty dictionaries can be written by following the type name of the dictionary by a pair of parenthesis, e.g. `[Int: String]()`, or, if the type can be determined from the context, as just a pair of square brackets surrounding a colon, `[:]`. Type names for dictionaries are written ...
+## Key Takeaways from this Exercise
```
I really like the rework of after.md.

stillleben commented on languages/swift/exercises/concept/dictionaries/.docs/introduction.md:
```diff
@@ -62,5 +62,17 @@ for (name, address) in addresses {
 
-[dictionaries]: https://docs.swift.org/swift-book/LanguageGuide/CollectionTypes.html#ID113
-[dictionary-docs]: https://developer.apple.com/documentation/swift/Dictionary
+Like the other collection types, dictionaries can be sorted according to arbitrary sorting functions. To do so, one must first define a function that takes two _key: value_ pairs, represented as tuples and returns a `Bool` such that the return value is `true` if the first _key: value_ pair should appear in the sorted result _before_ the second _key: value_ pair.
```
I appreciate that additional sorting section and it's well understandable for me.
